`TestCompletenessInit` race condition

## Failure

#### Link(s) to logs

- https://console.cloud.google.com/cloud-build/builds/162eff39-48cf-4b3d-aa0e-25fbdd06fd98;step=0?project=ci-account

#### Relevant snippet

```
    testing.go:1319: race detected during execution of test
--- FAIL: TestCompletenessInit (12.87s)
```

Took 10-15 minutes to reproduce locally, but I was able to trigger it.

```
==================
WARNING: DATA RACE
Read at 0x00c000702c60 by goroutine 37749:
  github.com/gravitational/trace.(*TraceErr).UserMessage()
      /Users/zmb/go/pkg/mod/github.com/gravitational/trace@v1.1.19/trace.go:291 +0xe4
  github.com/gravitational/trace.(*TraceErr).Error()
      /Users/zmb/go/pkg/mod/github.com/gravitational/trace@v1.1.19/trace.go:324 +0x28
  github.com/sirupsen/logrus.(*JSONFormatter).Format()
      /Users/zmb/go/pkg/mod/github.com/gravitational/logrus@v1.4.4-0.20210817004754-047e20245621/json_formatter.go:70 +0x184
  github.com/gravitational/teleport/lib/utils.(*JSONFormatter).Format()
      /Users/zmb/teleport/lib/utils/formatter.go:249 +0x1e0
  github.com/sirupsen/logrus.(*Entry).write()
      /Users/zmb/go/pkg/mod/github.com/gravitational/logrus@v1.4.4-0.20210817004754-047e20245621/entry.go:289 +0xe0
  github.com/sirupsen/logrus.(*Entry).log()
      /Users/zmb/go/pkg/mod/github.com/gravitational/logrus@v1.4.4-0.20210817004754-047e20245621/entry.go:252 +0x704
  github.com/sirupsen/logrus.(*Entry).Log()
      /Users/zmb/go/pkg/mod/github.com/gravitational/logrus@v1.4.4-0.20210817004754-047e20245621/entry.go:301 +0x80
  github.com/sirupsen/logrus.(*Entry).Warn()
      /Users/zmb/go/pkg/mod/github.com/gravitational/logrus@v1.4.4-0.20210817004754-047e20245621/entry.go:322 +0x29c
  github.com/gravitational/teleport/lib/cache.(*Cache).update()
      /Users/zmb/teleport/lib/cache/cache.go:823 +0x17c
  github.com/gravitational/teleport/lib/cache.(*Cache).Start.func1()
      /Users/zmb/teleport/lib/cache/cache.go:771 +0x64

Previous write at 0x00c000702c60 by goroutine 37460:
  github.com/gravitational/trace.(*TraceErr).AddUserMessage()
      /Users/zmb/go/pkg/mod/github.com/gravitational/trace@v1.1.19/trace.go:260 +0xe8
  github.com/gravitational/trace.Wrap()
      /Users/zmb/go/pkg/mod/github.com/gravitational/trace@v1.1.19/trace.go:64 +0x1d0
  github.com/gravitational/teleport/lib/cache.(*Cache).fetch.func2()
      /Users/zmb/teleport/lib/cache/cache.go:1262 +0x494
  golang.org/x/sync/errgroup.(*Group).Go.func1()
      /Users/zmb/go/pkg/mod/golang.org/x/sync@v0.1.0/errgroup/errgroup.go:75 +0x6c

Goroutine 37749 (running) created at:
  github.com/gravitational/teleport/lib/cache.(*Cache).Start()
      /Users/zmb/teleport/lib/cache/cache.go:771 +0x294
  github.com/gravitational/teleport/lib/cache.New()
      /Users/zmb/teleport/lib/cache/cache.go:749 +0x1524
  github.com/gravitational/teleport/lib/cache.TestCompletenessInit()
      /Users/zmb/teleport/lib/cache/cache_test.go:588 +0x68c
  testing.tRunner()
      /opt/homebrew/Cellar/go/1.19.4/libexec/src/testing/testing.go:1446 +0x188
  testing.(*T).Run.func1()
      /opt/homebrew/Cellar/go/1.19.4/libexec/src/testing/testing.go:1493 +0x40

Goroutine 37460 (finished) created at:
  golang.org/x/sync/errgroup.(*Group).Go()
      /Users/zmb/go/pkg/mod/golang.org/x/sync@v0.1.0/errgroup/errgroup.go:72 +0x10c
  github.com/gravitational/teleport/lib/cache.(*Cache).fetch()
      /Users/zmb/teleport/lib/cache/cache.go:1244 +0x520
  github.com/gravitational/teleport/lib/cache.(*Cache).fetchAndWatch()
      /Users/zmb/teleport/lib/cache/cache.go:941 +0x328
  github.com/gravitational/teleport/lib/cache.(*Cache).update()
      /Users/zmb/teleport/lib/cache/cache.go:817 +0x14c
  github.com/gravitational/teleport/lib/cache.(*Cache).Start.func1()
      /Users/zmb/teleport/lib/cache/cache.go:771 +0x64
==================
--- FAIL: TestCompletenessInit (2.47s)
    testing.go:1319: race detected during execution of test
FAIL
FAIL	github.com/gravitational/teleport/lib/cache	12.820s
FAIL
```

--- a/api/go.mod
+++ b/api/go.mod
@@ -6,7 +6,7 @@ require (
 	github.com/go-piv/piv-go v1.10.0
 	github.com/gogo/protobuf v1.3.2
 	github.com/google/go-cmp v0.5.9
-	github.com/gravitational/trace v1.1.19
+	github.com/gravitational/trace v1.1.20
 	github.com/jonboulle/clockwork v0.3.0
 	github.com/russellhaering/gosaml2 v0.8.1
 	github.com/sirupsen/logrus v1.9.0

--- a/api/go.mod
+++ b/api/go.mod
@@ -6,7 +6,7 @@ require (
 	github.com/go-piv/piv-go v1.10.0
 	github.com/gogo/protobuf v1.3.2
 	github.com/google/go-cmp v0.5.9
-	github.com/gravitational/trace v1.1.20
+	github.com/gravitational/trace v1.1.21-0.20221221165423-040ad96641b1
 	github.com/jonboulle/clockwork v0.3.0
 	github.com/russellhaering/gosaml2 v0.8.1
 	github.com/sirupsen/logrus v1.9.0

--- a/api/go.mod
+++ b/api/go.mod
@@ -6,7 +6,7 @@ require (
 	github.com/go-piv/piv-go v1.10.0
 	github.com/gogo/protobuf v1.3.2
 	github.com/google/go-cmp v0.5.9
-	github.com/gravitational/trace v1.1.21-0.20221221165423-040ad96641b1
+	github.com/gravitational/trace v1.2.0
 	github.com/jonboulle/clockwork v0.3.0
 	github.com/russellhaering/gosaml2 v0.8.1
 	github.com/sirupsen/logrus v1.9.0

--- a/go.mod
+++ b/go.mod
@@ -72,7 +72,7 @@ require (
 	github.com/gravitational/reporting v0.0.0-20210923183620-237377721140
 	github.com/gravitational/roundtrip v1.0.2
 	github.com/gravitational/teleport/api v0.0.0
-	github.com/gravitational/trace v1.1.20
+	github.com/gravitational/trace v1.1.21-0.20221221165423-040ad96641b1
 	github.com/gravitational/ttlmap v0.0.0-20171116003245-91fd36b9004c
 	github.com/grpc-ecosystem/go-grpc-middleware/providers/openmetrics/v2 v2.0.0-rc.3
 	github.com/hashicorp/golang-lru v0.6.0

--- a/go.mod
+++ b/go.mod
@@ -71,6 +71,7 @@ require (
 	github.com/gravitational/oxy v0.0.0-20221029012416-9fbf4c444680
 	github.com/gravitational/reporting v0.0.0-20210923183620-237377721140
 	github.com/gravitational/roundtrip v1.0.2
+	github.com/gravitational/teleport/api v0.0.0
 	github.com/gravitational/trace v1.2.0
 	github.com/gravitational/ttlmap v0.0.0-20171116003245-91fd36b9004c
 	github.com/grpc-ecosystem/go-grpc-middleware/providers/openmetrics/v2 v2.0.0-rc.3

--- a/go.mod
+++ b/go.mod
@@ -71,8 +71,7 @@ require (
 	github.com/gravitational/oxy v0.0.0-20221029012416-9fbf4c444680
 	github.com/gravitational/reporting v0.0.0-20210923183620-237377721140
 	github.com/gravitational/roundtrip v1.0.2
-	github.com/gravitational/teleport/api v0.0.0
-	github.com/gravitational/trace v1.1.21-0.20221221165423-040ad96641b1
+	github.com/gravitational/trace v1.2.0
 	github.com/gravitational/ttlmap v0.0.0-20171116003245-91fd36b9004c
 	github.com/grpc-ecosystem/go-grpc-middleware/providers/openmetrics/v2 v2.0.0-rc.3
 	github.com/hashicorp/golang-lru v0.6.0

--- a/go.mod
+++ b/go.mod
@@ -72,7 +72,7 @@ require (
 	github.com/gravitational/reporting v0.0.0-20210923183620-237377721140
 	github.com/gravitational/roundtrip v1.0.2
 	github.com/gravitational/teleport/api v0.0.0
-	github.com/gravitational/trace v1.1.19
+	github.com/gravitational/trace v1.1.19-0.20221215085259-16163cb74632
 	github.com/gravitational/ttlmap v0.0.0-20171116003245-91fd36b9004c
 	github.com/grpc-ecosystem/go-grpc-middleware/providers/openmetrics/v2 v2.0.0-rc.3
 	github.com/hashicorp/golang-lru v0.6.0

--- a/go.mod
+++ b/go.mod
@@ -71,7 +71,6 @@ require (
 	github.com/gravitational/oxy v0.0.0-20221029012416-9fbf4c444680
 	github.com/gravitational/reporting v0.0.0-20210923183620-237377721140
 	github.com/gravitational/roundtrip v1.0.2
-	github.com/gravitational/teleport/api v0.0.0
 	github.com/gravitational/trace v1.1.19-0.20221215085259-16163cb74632
 	github.com/gravitational/ttlmap v0.0.0-20171116003245-91fd36b9004c
 	github.com/grpc-ecosystem/go-grpc-middleware/providers/openmetrics/v2 v2.0.0-rc.3

--- a/go.mod
+++ b/go.mod
@@ -71,7 +71,8 @@ require (
 	github.com/gravitational/oxy v0.0.0-20221029012416-9fbf4c444680
 	github.com/gravitational/reporting v0.0.0-20210923183620-237377721140
 	github.com/gravitational/roundtrip v1.0.2
-	github.com/gravitational/trace v1.1.19-0.20221215085259-16163cb74632
+	github.com/gravitational/teleport/api v0.0.0
+	github.com/gravitational/trace v1.1.20
 	github.com/gravitational/ttlmap v0.0.0-20171116003245-91fd36b9004c
 	github.com/grpc-ecosystem/go-grpc-middleware/providers/openmetrics/v2 v2.0.0-rc.3
 	github.com/hashicorp/golang-lru v0.6.0

--- a/lib/auth/auth.go
+++ b/lib/auth/auth.go
@@ -1856,16 +1856,14 @@ func (a *Server) WithUserLock(username string, authenticateFn func() error) erro
 				user.GetName(), defaults.MaxAccountRecoveryAttempts, apiutils.HumanTimeFormat(status.RecoveryAttemptLockExpires))
 
 			err := trace.AccessDenied(MaxFailedAttemptsErrMsg)
-			err.AddField(ErrFieldKeyUserMaxedAttempts, true)
-			return err
+			return trace.WithField(err, ErrFieldKeyUserMaxedAttempts, true)
 		}
 		if status.LockExpires.After(a.clock.Now().UTC()) {
 			log.Debugf("%v exceeds %v failed login attempts, locked until %v",
 				user.GetName(), defaults.MaxLoginAttempts, apiutils.HumanTimeFormat(status.LockExpires))
 
 			err := trace.AccessDenied(MaxFailedAttemptsErrMsg)
-			err.AddField(ErrFieldKeyUserMaxedAttempts, true)
-			return err
+			return trace.WithField(err, ErrFieldKeyUserMaxedAttempts, true)
 		}
 	}
 	fnErr := authenticateFn()
@@ -1909,8 +1907,7 @@ func (a *Server) WithUserLock(username string, authenticateFn func() error) erro
 	}
 
 	retErr := trace.AccessDenied(MaxFailedAttemptsErrMsg)
-	retErr.AddField(ErrFieldKeyUserMaxedAttempts, true)
-	return retErr
+	return trace.WithField(retErr, ErrFieldKeyUserMaxedAttempts, true)
 }
 
 // PreAuthenticatedSignIn is for MFA authentication methods where the password

--- a/lib/auth/github.go
+++ b/lib/auth/github.go
@@ -504,19 +504,20 @@ func (a *Server) validateGithubAuthCallback(ctx context.Context, diagCtx *SSODia
 
 		// optional parameter: error_description
 		errDesc := q.Get("error_description")
-		return nil, trace.OAuth2(oauth2.ErrorInvalidRequest, errParam, q).AddUserMessage("Github returned error: %v [%v]", errDesc, errParam)
+		oauthErr := trace.OAuth2(oauth2.ErrorInvalidRequest, errParam, q)
+		return nil, trace.WithUserMessage(oauthErr, "Github returned error: %v [%v]", errDesc, errParam)
 	}
 
 	code := q.Get("code")
 	if code == "" {
-		return nil, trace.OAuth2(oauth2.ErrorInvalidRequest,
-			"code query param must be set", q).AddUserMessage("Invalid parameters received from Github.")
+		oauthErr := trace.OAuth2(oauth2.ErrorInvalidRequest, "code query param must be set", q)
+		return nil, trace.WithUserMessage(oauthErr, "Invalid parameters received from Github.")
 	}
 
 	stateToken := q.Get("state")
 	if stateToken == "" {
-		return nil, trace.OAuth2(oauth2.ErrorInvalidRequest,
-			"missing state query param", q).AddUserMessage("Invalid parameters received from Github.")
+		oauthErr := trace.OAuth2(oauth2.ErrorInvalidRequest, "missing state query param", q)
+		return nil, trace.WithUserMessage(oauthErr, "Invalid parameters received from Github.")
 	}
 	diagCtx.RequestID = stateToken
 

--- a/lib/services/lock.go
+++ b/lib/services/lock.go
@@ -35,8 +35,7 @@ func LockInForceAccessDenied(lock types.Lock) error {
 		s += ": " + msg
 	}
 	err := trace.AccessDenied(s)
-	err.AddField("lock-in-force", lock)
-	return err
+	return trace.WithField(err, "lock-in-force", lock)
 }
 
 // StrictLockingModeAccessDenied is an AccessDenied error returned when strict

